### PR TITLE
Refactor pointer events to use geometry types

### PR DESCRIPTION
**Pull request by wmww**
Also adds a single pointer event builder method that includes all properties, and deprecates the older event builder methods.

### CODE REVIEW

**Comment by wmww** [comment]
I know `mir::geometry::generic::Displacement<mir::geometry::generic::Value<bool>::Wrapper>` is quite ugly, but it will collapse down into `mir::geometry::generic::Displacement<bool>` once we get #2475 landed.

**Comment by AlanGriffiths** [comment]
> once we get #2475 landed

At the current rate of progress that could be some time! :wink:

**Review by RAOF — COMMENTED** [review]
I'm not sure what I think this should *eventually* look like, but I think this is an improvement on the status quo.

I *think* that scroll would be clearer if it were not handled as a `Displacement`, but I'm willing to hear arguments :wink:

**Comment by AlanGriffiths** [comment]
> I _think_ that scroll would be clearer if it were not handled as a `Displacement`

I concur. scroll events are handled as horizontal/vertical scrolls not like move events

**Comment by wmww** [comment]
> They're logically two separate event-like-things rather than a single logical event like motion

> scroll events are handled as horizontal/vertical scrolls not like move events

Either I'm not understanding or we have a different impression of what scroll is. To me, they're just a different type of 2D motion. The Wayland protocol has this to say:

> For scroll events (vertical and horizontal scroll axes), the value parameter is the length of a vector along the specified axis in a coordinate space identical to those of motion events, representing a relative movement along the specified axis.

I see how `Displacement<bool>` is a little strange. It's also a little strange to allow a non-zero scroll value and also `scroll_stop`, but I don't think particularly harmful. The variant approach would arguably be more elegant, but I'm not sure it would yield any significant practical benefit.

**Comment by wmww** [comment]
After discussions with the team, we've decided to have a `ScrollAxis` struct, and then have horizontal and vertical instances of that.

File: include/common/mir/events/event_builders.h
Comment (RAOF) [inline include/common/mir/events/event_builders.h:113]: Some places you use `ScrollAxisV1<Tag>`, some places you use `ScrollAxis<Tag>`, and some places you use `ScrollAxisH`/`ScrollAxisV`. There's presumably some reason behind this, but I can't see it?

**Review by RAOF — COMMENTED** [review]
Yeah, that's nicer!

I *think* we should be using `ScrollAxisV`/`ScrollAxisH` everywhere, though? I don't *think* there's any reason not to? The underlying type is `ScrollAxisV1<Tag>`, so we can still add a `ScrollAxisV2<Tag>` without breaking ABI, and I *think* that we can even do that more easily if the API is the `ScrollAxisV/H` types, as then you might not need source changes to update to the new stuff?

File: include/platform/mir/input/event_builder.h
Comment (AlanGriffiths) [inline include/platform/mir/input/event_builder.h:58]: Is there any point to retaining these APIs?

File: include/common/mir/events/event_builders.h
Comment (wmww) [reply include/common/mir/events/event_builders.h:113]: How it should be is `ScrollAxisV1<Tag>` is used in places where we don't want the version auto-updated (for example, in function signatures of ABI-stable functions), `ScrollAxis<Tag>` is used in places that handle the tag generically but should always use the latest version and `ScrollAxisH`/`ScrollAxisV` are used most places where neither of those are the case. As it is now there are residual uses of `ScrollAxis<Tag>`, which I'll fix.

File: include/platform/mir/input/event_builder.h
Comment (wmww) [reply include/platform/mir/input/event_builder.h:58]: As far as I know we're not breaking platform ABI

File: src/common/events/event_builders.cpp
Comment (RAOF) [inline src/common/events/event_builders.cpp:253]: I think it's important for the next person who's reading this to document that this is *deliberately* using the `V1` types while the header declares these parameters with the unversioned alias so that we get a compile error should someone bump the unversioned alias without remembering to do the rest of the work to maintain ABI.

**Review by RAOF — APPROVED** [review]
LGTM, modulo the addition of the comment.

bors delegate=wmww

File: include/platform/mir/input/event_builder.h
Comment (AlanGriffiths) [reply include/platform/mir/input/event_builder.h:58]: This isn't an ABI we're expecting outside folks to use yet. So breaking it isn't a big deal

**Pull request merged**
